graphpatch/extraction/extraction_context.py

```diff
@@ -0,0 +1,251 @@
+from contextlib import ExitStack, contextmanager
+from copy import deepcopy
+from enum import Enum
+from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple, Type, Union
+
+import torch
+from torch.fx import Graph
+from torch.nn import LayerNorm, Module, ModuleDict, ModuleList
+
+from .. import hacks
+from ..optional.accelerate import ModelHook
+from ..optional.bitsandbytes import Linear8bitLt
+from ..optional.typing_extensions import TypeAlias, TypeGuard
+from .graphpatch_module import GraphPatchModule
+from .wrapped_8_bit_linear import Wrapped8BitLinear
+from .wrapped_layer_norm import WrappedLayerNorm
+
+CONTAINER_TYPES = (ModuleList, ModuleDict)
+WrappedModule: TypeAlias = Union["ExtractionWrapper", ModuleDict, ModuleList]
+ExtractionMethod = Enum("ExtractionMethod", ("custom", "compiled", "opaque"))
+
+
+def init_container(container: Union[ModuleList, ModuleDict]) -> Union[ModuleList, ModuleDict]:
+    if isinstance(container, ModuleList):
+        return type(container)([Module() for _ in range(len(container))])
+    return type(container)()
+
+
+def is_container(module: Union[Module, Type[Module]]) -> TypeGuard[Union[ModuleList, ModuleDict]]:
+    """Strictly checking for built-in container types, since user-derived ones could have forward()."""
+    if isinstance(module, type):
+        model_class = module
+    else:
+        model_class = type(module)
+    return model_class in CONTAINER_TYPES
+
+
+class ModuleInvocation:
+    args: Tuple[Any]
+    kwargs: Dict[str, Any]
+    output: Any
+
+    @hacks.skip  # type: ignore
+    def __init__(
+        self,
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+        output: Any,
+    ):
+        self.args = args
+        self.kwargs = kwargs
+        self.output = output
+
+
+class ExtractionState:
```

Comment:
I'm freezing refactors for this update, but I think this should probably be rolled in to `ExtractionWrapper` since they are one-to-one anyway.